test/integration/targets/module_defaults/test_defaults.yml

```diff
@@ -105,7 +105,7 @@
         - "legacy_ping_2.ping == 'from task'"
         - "legacy_ping_3.ping == 'from task'"
         - "legacy_ping_4.ping == 'from task'"
-        - "legacy_builtin_defaults.msg == 'Hello world!'"
+        - "legacy_builtin_defaults.msg == 'legacy default'"
```

Comment:
Changing this behavior is a bug, imo because the action doesn't resolve to ansible.builtin.debug, it's executing a plugin in a legacy path https://github.com/ansible/ansible/blob/devel/test/integration/targets/module_defaults/action_plugins/debug.py. ansible.legacy is a superset of ansible.builtin, so ansible.legacy defaults can apply to ansible.builtin tasks, but not vice-versa.

I liked your original fix for the action plugins, for the (normal) module issue we could just patch the action plugin:

<details>
<summary>diff</summary>

```diff
diff --git a/lib/ansible/plugins/action/__init__.py b/lib/ansible/plugins/action/__init__.py
index 9f7fac1b0b..13c383eaab 100644
--- a/lib/ansible/plugins/action/__init__.py
+++ b/lib/ansible/plugins/action/__init__.py
@@ -21,6 +21,7 @@ from ansible import constants as C
 from ansible.errors import AnsibleError, AnsibleConnectionFailure, AnsibleActionSkip, AnsibleActionFail, AnsibleAuthenticationFailure
 from ansible.executor.module_common import modify_module
 from ansible.executor.interpreter_discovery import discover_interpreter, InterpreterDiscoveryRequiredError
+from ansible.executor.module_common import get_action_args_with_defaults
 from ansible.module_utils.common.arg_spec import ArgumentSpecValidator
 from ansible.module_utils.errors import UnsupportedError
 from ansible.module_utils.json_utils import _filter_non_json_lines
@@ -68,6 +69,8 @@ class ActionBase(ABC):
     _supports_check_mode = True
     _supports_async = False
+    _load_defaults_on_execute = False
+
     def __init__(self, task, connection, play_context, loader, templar, shared_loader_obj):
         self._task = task
         self._connection = connection
@@ -277,6 +280,13 @@ class ActionBase(ABC):
         else:  # This is a for-else: http://bit.ly/1ElPkyg
             raise AnsibleError("The module %s was not found in configured module paths" % (module_name))
 
+        # load resolved module_defaults
+        if self._load_defaults_on_execute:
+            module_args = get_action_args_with_defaults(
+                result.resolved_fqcn, module_args, self._task.module_defaults,
+                self._templar, action_groups=self._task._parent._play._action_groups,
+            )
+
         # insert shared code and arguments into the module
         final_environment = dict()
         self._compute_environment_string(final_environment)

diff --git a/lib/ansible/plugins/action/normal.py b/lib/ansible/plugins/action/normal.py
index 0476f9aacd..8af3931b3e 100644
--- a/lib/ansible/plugins/action/normal.py
+++ b/lib/ansible/plugins/action/normal.py
@@ -26,6 +26,8 @@ class ActionModule(ActionBase):
     _supports_check_mode = True
     _supports_async = True
 
+    _load_defaults_on_execute = True
+
     def run(self, tmp=None, task_vars=None):
 
         # individual modules might disagree but as the generic the action plugin, pass at this point.
```
</details>

That would have the added benefit of letting us simplify existing action plugins too (tested in branch https://github.com/ansible/ansible/pull/74103/commits/266483cd2d56b13e381c3e88b16cfbc01546a5a0, nothing seems to blow up).